go-controller/pkg/ovn/egressgw.go

```diff
@@ -0,0 +1,254 @@
+	}
+	routingNamespaceAnnotation := pod.Annotations[routingNamespaceAnnotation]
+	if routingNamespaceAnnotation == "" {
+		return nil
+	}
+	klog.Infof("External gateway pod: %s, detected for namespace(s) %s", pod.Name, routingNamespaceAnnotation)
+	var foundGws []net.IP
+	if pod.Spec.HostNetwork {
+		for _, podIP := range pod.Status.PodIPs {
+			ip := net.ParseIP(podIP.IP)
+			if ip != nil {
+				foundGws = append(foundGws, ip)
+			}
+		}
+	} else if pod.Annotations[routingNetworkAnnotation] != "" {
+		var multusNetworks []Network
+		err := json.Unmarshal([]byte(pod.ObjectMeta.Annotations["k8s.v1.cni.cncf.io/network-status"]), &multusNetworks)
+		if err != nil {
+			return fmt.Errorf("unable to unmarshall annotation k8s.v1.cni.cncf.io/network-status on pod %s: %v", pod.Name, err)
+		}
+		for _, multusNetwork := range multusNetworks {
+			if multusNetwork.Name == routingNetworkAnnotation {
+				for _, gwIP := range multusNetwork.Ips {
+					ip := net.ParseIP(gwIP)
+					if ip != nil {
+						foundGws = append(foundGws, ip)
+					}
+				}
+			}
+		}
+	} else {
+		klog.Errorf("Ignoring pod %s as an external gateway candidate. Invalid combination "+
+			"of host network: %t and routing-network annotation: %s", pod.Name, pod.Spec.HostNetwork, routingNetworkAnnotation)
+		return nil
+	}
+
+	// if we found any gateways then we need to update current pods routing in the relevant namespace
+	if len(foundGws) == 0 {
+		klog.Warningf("No valid gateway IPs found for requested external gateway pod: %s", pod.Name)
+		return nil
+	}
+
+	for _, namespace := range strings.Split(routingNamespaceAnnotation, ",") {
+		err := oc.addPodExternalGWForNamespace(namespace, pod, foundGws)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addPodExternalGWForNamespace handles adding routes to all pods in that namespace for a pod GW
+func (oc *Controller) addPodExternalGWForNamespace(namespace string, pod *kapi.Pod, gws []net.IP) error {
+	nsInfo, err := oc.waitForNamespaceLocked(namespace)
+	if err != nil {
+		return err
+	}
+	defer nsInfo.Unlock()
+	nsInfo.routingExternalPodGWs[pod.Name] = gws
+	return oc.addGWRoutesForNamespace(namespace, gws, nsInfo)
+}
+
+// addExternalGWsForNamespace handles adding annotated gw routes to all pods in namespace
+// This should only be called with a lock on nsInfo
+func (oc *Controller) addExternalGWsForNamespace(gateways []net.IP, nsInfo *namespaceInfo, namespace string) error {
+	if gateways == nil {
+		return fmt.Errorf("unable to add gateways routes for namespace: %s, gateways are nil", namespace)
+	}
+	nsInfo.routingExternalGWs = gateways
+	return oc.addGWRoutesForNamespace(namespace, gateways, nsInfo)
+}
+
+// addGWRoutesForNamespace handles adding routes for all existing pods in namespace
+// This should only be called with a lock on nsInfo
+func (oc *Controller) addGWRoutesForNamespace(namespace string, gws []net.IP, nsInfo *namespaceInfo) error {
+	existingPods, err := oc.watchFactory.GetPods(namespace)
```

Comment:
do we want to get ALL pods, only the running pods or pods that are not `pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed` ?

Reply:
hmm good question. This is for namespace update, so we definitely need to cover all pods that have succeeded. If a pod is failed, I guess it will retry again later in the pod watcher and then would get handled there to have its route updated. Is that your thought process as well?

Reply:
meanwhile we are consistent I don't think there should be a problem, I was worried if we delete on a Pod delete event and if we will try to delete again here or something similar

Reply:
yeah this is the consistent way we handle this with other namespace watching today (and some other handlers as well I believe). But we probably could potentially filter out some of the pods we don't care about. I think you could look at that holistically across the entire code base and make that change in another PR.